src/Microsoft.AspNetCore.Proxy/ProxyMiddleware.cs

```diff
@@ -59,8 +65,70 @@ public ProxyMiddleware(RequestDelegate next, IOptions<ProxyOptions> options)
             _httpClient = new HttpClient(_options.BackChannelMessageHandler ?? new HttpClientHandler());
         }
 
-        public async Task Invoke(HttpContext context)
+        public Task Invoke(HttpContext context) => context.WebSockets.IsWebSocketRequest ? HandleWebSocketRequest(context) : HandleHttpRequest(context);
+
+        private async Task HandleWebSocketRequest(HttpContext context)
+        {
+            using (var client = new ClientWebSocket())
+            {
+                foreach (var headerEntry in context.Request.Headers)
+                {
+                    if (!NotForwardedWebSocketHeaders.Contains(headerEntry.Key, StringComparer.OrdinalIgnoreCase))
+                    {
+                        client.Options.SetRequestHeader(headerEntry.Key, headerEntry.Value);
+                    }
+                }
+
+                var wsScheme = string.Equals(_options.Scheme, "https", StringComparison.OrdinalIgnoreCase) ? "wss" : "ws";
+                var uriString = $"{wsScheme}://{_options.Host}:{_options.Port}{context.Request.PathBase}{context.Request.Path}{context.Request.QueryString}";
+
+                if (_options.WebSocketKeepAliveInterval.HasValue)
+                {
+                    client.Options.KeepAliveInterval = _options.WebSocketKeepAliveInterval.Value;
+                }
+
+                try
+                {
+                    await client.ConnectAsync(new Uri(uriString), context.RequestAborted);
+                }
+                catch (WebSocketException ex)
+                {
+                    HandleWebSocketConnectError(context, ex);
+                    return;
+                }
+
+                using (var server = await context.WebSockets.AcceptWebSocketAsync(client.SubProtocol))
+                {
+                    await Task.WhenAll(PumpWebSocket(client, server, context.RequestAborted), PumpWebSocket(server, client, context.RequestAborted));
+                }
+            }
+        }
+
+        private void HandleWebSocketConnectError(HttpContext context, WebSocketException ex)
         {
+            context.Response.StatusCode = 400;
+        }
+
+        private async Task PumpWebSocket(WebSocket source, WebSocket dest, CancellationToken ct)
+        {
+            var buffer = new byte[_options.WebSocketBufferSize ?? DefaultWebSocketBufferSize];
+            while (true)
+            {
+                var res = await source.ReceiveAsync(new ArraySegment<byte>(buffer), ct);
```

Comment:
I partially agree: For receive error there should be websocket closure with InternalServerError instead of 500.
I have pushed fixes (I also have renamed some more variables to match your conventions).